# Segmenting and Clustering Neighborhoods in Toronto

###### Author: Adam Kassoff

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis


## Part 1:
1. Creat notebook
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below: <img src='https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1603324800000&hmac=WmqqbK-twXsXwEaCPFbaPEoGST65EIKqAcXdBRU0nqQ' >

- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
- Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [4]:
wiki_data = pd.read_html(wiki_url.content)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We only care about the first table, so lets drop the other tables.

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Now we need to drop the data with not assigned values for borough.

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df.reset_index(inplace = True)
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Now let's group the data by the Postal Codes

In [8]:
df.groupby(['Postal Code']).first()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

Remove the redundant "index" column

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
df.shape

(103, 3)

### Answer to Part 1: We have 103 rows and 3 columns

________________________________

# Part 2

- Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.
- Use the Geocoder package or the csv file to create the following dataframe:
<img src='https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/HZ3jNHNOEeiMwApe4i-fLg_f44f0f10ccfaf42fcbdba9813364e173_Screen-Shot-2018-06-18-at-7.18.16-PM.png?expiry=1603324800000&hmac=pAe0lBDXVwV8osg8Q1YXCqKpMKtDWT0wt2myXP1-99k'>

Before we import geocoder we have to install geocoder

In [12]:
pip install geocoder

     |████████████████████████████████| 98 kB 6.8 MB/s eta 0:00:011
Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder

The following code didn't work and seemed to be stuck in an infinte loop.
>#initialize your variable to None <br>
lat_lng_coords = None
>
>postal_code = 'MG5'
>
>#loop until you get the coordinates <br>
>while(lat_lng_coords is None): <br>
>  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code)) <br>
>  lat_lng_coords = g.latlng <br>
>
>latitude = lat_lng_coords[0] <br>
longitude = lat_lng_coords[1]

#### Let's try using the csv file

In [14]:
df_geo = pd.read_csv("https://cocl.us/Geospatial_data")
df_geo.head

<bound method NDFrame.head of     Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]>

Let's check the data types of our columns specifically the postal codes, so we can combine the two data frames

In [15]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

Let's also check that both the shapes match so no postal codes are missing

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", df_geo.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


### Here is were we join both data frames to get our desired result for part 2

In [18]:
combined_data = df.join(df_geo.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [19]:
combined_data.shape

(103, 5)

______

# Part 3

#### Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

In [20]:
from geopy.geocoders import Nominatim

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Let's vizualize the map of Toronto

In [22]:
!pip install folium
import folium

     |████████████████████████████████| 93 kB 3.8 MB/s  eta 0:00:01


In [23]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Define Foursquare API Credentials and Version

In [25]:
CLIENT_ID = 'FFN1U3J0RP1BGN41N50XKRUTNJPJQ1WH1Y1SSPGTB5M3Y3F2' # your Foursquare ID
CLIENT_SECRET = 'QBRMF3YWOD2OO1W1ZWISMUXGWZ4WQYBABNRT5DGCBFTCFH23' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FFN1U3J0RP1BGN41N50XKRUTNJPJQ1WH1Y1SSPGTB5M3Y3F2
CLIENT_SECRET:QBRMF3YWOD2OO1W1ZWISMUXGWZ4WQYBABNRT5DGCBFTCFH23


Let's explore the data, and get the venues in 500 meters range from our first entry

In [32]:
neighbourhood_latitude = combined_data.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = combined_data.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = combined_data.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Create the GET request URL

In [33]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)Create the GET request URL
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=FFN1U3J0RP1BGN41N50XKRUTNJPJQ1WH1Y1SSPGTB5M3Y3F2&client_secret=QBRMF3YWOD2OO1W1ZWISMUXGWZ4WQYBABNRT5DGCBFTCFH23&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f8f5c2c349fac2158b6268d'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [35]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [36]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [37]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [40]:
toronto_venues = getNearbyVenues(names=combined_data['Neighbourhood'],
                                   latitudes=combined_data['Latitude'],
                                   longitudes=combined_data['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Checking the size of df

In [41]:
print(toronto_venues.shape)
toronto_venues.head()

(2136, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

Let's check how many venues were returned for each neighborhood

In [42]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                 7   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            5   
Woburn                                                                 4   
Woodbine Heights                                                       7   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                

In [43]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 273 uniques categories.


Analyze each neighbourhood

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 274 columns]

In [45]:
toronto_onehot.shape

(2136, 274)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [46]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
91                      Willowdale, Willowdale West                0.0   
92                                           Woburn                0.0   
93                                 Woodbine Heights                0.0   
94                                  York Mills West                0.0   
95                         York Mills, Silver Hills                0.0   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                 0.0      0.0                 0.0           0.0   
1                 0.0      0.0                 0.0           0.0   
2                 0.0      0.0                 0.0           0.0   
3                 0.0      0.0                 0.0           0.0   
4                 0.0      0.0                 0.0           0.0   
..                ...      ...                 ...           ...   
91                0.0      0.0                 0.0           0.0   
92                0.0      0.0                 0.0           0.0   
93                0.0      0.0                 0.0           0.0   
94                0.0      0.0                 0.0           0.0   
95                0.0      0.0                 0.0           0.0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0              0.0              0.0               0.0             0.000000   
1              0.0              0.0               0.0             0.000000   
2              0.0              0.0               0.0             0.000000   
3              0.0              0.0               0.0             0.000000   
4              0.0              0.0               0.0             0.045455   
..             ...              ...               ...                  ...   
91             0.0              0.0               0.0             0.000000   
92             0.0              0.0               0.0             0.000000   
93             0.0              0.0               0.0             0.000000   
94             0.0              0.0               0.0             0.000000   
95             0.0              0.0               0.0             0.000000   

    ...  Train Station  Turkish Restaurant  Vegetarian / Vegan Restaurant  \
0   ...            0.0                 0.0                            0.0   
1   ...            0.0                 0.0                            0.0   
2   ...            0.0                 0.0                            0.0   
3   ...            0.0                 0.0                            0.0   
4   ...            0.0                 0.0                            0.0   
..  ...            ...                 ...                            ...   
91  ...            0.0                 0.0                            0.0   
92  ...            0.0                 0.0                            0.0   
93  ...            0.0                 0.0                            0.0   
94  ...            0.0                 0.0                            0.0   
95  ...            0.0                 0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0                    0.0              0.0       0.0   
1                0.0                    0.0              0.0       0.0   
2                0.0                    0.0              0.0       0.0   
3                0.0                    0.0              0.0       0.0   
4                0.0                    0.0              0.0       0.0   
..               ...      

In [48]:
toronto_grouped.shape

(96, 274)

Getting the top 10 for each neighbourhood

In [64]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [65]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Clothing Store   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot          Skating Rink   
1        Sandwich Place           Coffee Shop                   Pub   
2                  Bank              Pharmacy        Ice Cream Shop   
3                  Café    Chinese Restaurant                  Bank   
4    Italian Restaurant           Coffee Shop      Greek Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Latin American Restaurant     Electronics Store      Doner Restaurant   
1                   Pharmacy                   Gym      Greek Restaurant   
2                Bridal Shop         Shopping Mall        Sandwich Place   
3        Distribution Center               Dog Run      Doner Restaurant   
4            Thai Restaurant             Locksmith          Liquor Store   

     8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0               Donut Shop                  Drugstore    Dumpling Restaurant  
1           Discount Store           Department Store           Dessert Shop  
2                    Diner  Middle Eastern Restaurant             Restaurant  
3               Donut Shop                  Drugstore            Yoga Studio  
4  Comfort Food Restaurant                  Juice Bar                Butcher

#### Clustering k=5

In [66]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [67]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = combined_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio    Dumpling Restaurant   
1          Hockey Arena           Coffee Shop  Portuguese Restaurant   
2                Bakery                   Pub                   Park   
3     Accessories Store              Boutique              Gift Shop   
4           Yoga Studio    College Auditorium                    Bar   

    5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
0     Distribution Center                     Dog Run      Doner Restaurant   
1            Intersection  Construction & Landscaping    Dim Sum Restaurant   
2          Breakfast Spot                        Café               Theater   
3  Furniture / Home Store                 Event Space           Coffee Shop   
4                Beer Bar               Smoothie Shop        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Donut Shop              Drugstore      Electronics Store  
1           Event Space   Ethiopian Restaurant            Escape Room  
2                   Spa                Brewery             Shoe Store  
3         Women's Store  Vietnamese Restaurant        Airport Service  
4                  Café             Restaurant     Chinese Restaurant

Check for NaN values and drop them to prevent data skew

In [68]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough                            Neighbourhood  \
5          M9A    Etobicoke  Islington Avenue, Humber Valley Village   
52         M2M   North York                  Willowdale, Newtonbrook   
95         M1X  Scarborough                              Upper Rouge   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
5   43.667856 -79.532242             NaN                   NaN   
52  43.789053 -79.408493             NaN                   NaN   
95  43.836125 -79.205636             NaN                   NaN   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                    NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5                    NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                    NaN                   NaN                    NaN  
52                   NaN                   NaN                    NaN  
95                   NaN                   NaN                    NaN

In [69]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plot the clusters in the map

In [70]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [74]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Examine Clusters

#### Cluster 1

In [75]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0          Intersection   
49        North York             0.0                Bakery   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue        3rd Most Common Venue  \
0      Food & Drink Shop                  Yoga Studio   
21         Women's Store                         Pool   
35                  Park            Convenience Store   
49                  Park   Construction & Landscaping   
61           Swim School                     Bus Line   
64           Yoga Studio  Eastern European Restaurant   
66     Convenience Store                  Yoga Studio   
85                  Park                       Bakery   
91            Playground                        Trail   

          4th Most Common Venue        5th Most Common Venue  \
0           Dumpling Restaurant          Distribution Center   
21                  Yoga Studio          Dumpling Restaurant   
35          Dumpling Restaurant               Discount Store   
49                  Yoga Studio  Eastern European Restaurant   
61                  Yoga Studio                    Drugstore   
64          Distribution Center                      Dog Run   
66  Eastern European Restaurant          Distribution Center   
85                  Yoga Studio          Dumpling Restaurant   
91                  Yoga Studio                    Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Dog Run      Doner Restaurant            Donut Shop   
21        Discount Store   Distribution Center               Dog Run   
35   Distribution Center               Dog Run      Doner Restaurant   
49               Dog Run      Doner Restaurant            Donut Shop   
61   Distribution Center               Dog Run      Doner Restaurant   
64      Doner Restaurant            Donut Shop             Drugstore   
66               Dog Run      Doner Restaurant            Donut Shop   
85   Distribution Center               Dog Run      Doner Restaurant   
91                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
0              Drugstore      Electronics Store  
21      Doner Restaurant             Donut Shop  
35            Donut Shop              Drugstore  
49             Drugstore    Dumpling Restaurant  
61            Donut Shop    Dumpling Restaurant  
64   Dumpling Restaurant      Electronics Store  
66             Drugstore    Dumpling Restaurant  
85            Donut Shop              Drugstore  
91               Dog Run       Doner Restaurant

#### Cluster 2

In [76]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York             1.0                 Pizza Place   
2    Downtown Toronto             1.0                 Coffee Shop   
3          North York             1.0              Clothing Store   
4    Downtown Toronto             1.0                 Coffee Shop   
7          North York             1.0                         Gym   
..                ...             ...                         ...   
98          Etobicoke             1.0                       River   
99   Downtown Toronto             1.0                 Coffee Shop   
100      East Toronto             1.0                 Pizza Place   
101         Etobicoke             1.0  Construction & Landscaping   
102         Etobicoke             1.0                         Gym   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
1            Hockey Arena           Coffee Shop        Portuguese Restaurant   
2                  Bakery                   Pub                         Park   
3       Accessories Store              Boutique                    Gift Shop   
4             Yoga Studio    College Auditorium                          Bar   
7     Japanese Restaurant            Beer Store                  Coffee Shop   
..                    ...                   ...                          ...   
98                   Pool           Yoga Studio                   Donut Shop   
99                Gay Bar   Japanese Restaurant             Sushi Restaurant   
100         Auto Workshop         Garden Center                       Garden   
101        Baseball Field           Yoga Studio  Eastern European Restaurant   
102           Social Club        Discount Store                       Bakery   

      5th Most Common Venue       6th Most Common Venue  \
1              Intersection  Construction & Landscaping   
2            Breakfast Spot                        Café   
3    Furniture / Home Store                 Event Space   
4                  Beer Bar               Smoothie Shop   
7      Caribbean Restaurant              Clothing Store   
..                      ...                         ...   
98                    Diner              Discount Store   
99               Restaurant                 Yoga Studio   
100      Light Rail Station        Fast Food Restaurant   
101                 Dog Run            Doner Restaurant   
102           Tanning Salon               Burrito Place   

        7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
1          Dim Sum Restaurant           Event Space   Ethiopian Restaurant   
2                     Theater                   Spa                Brewery   
3                 Coffee Shop         Women's Store  Vietnamese Restaurant   
4              Sandwich Place                  Café             Restaurant   
7          Italian Restaurant        Discount Store                   Café   
..                        ...                   ...                    ...   
98        Distribution Center               Dog Run       Doner Restaurant   
99   Mediterranean Restaurant                  Café                    Pub   
100            Farmers Market            Comic Shop                   Park   
101                Donut Shop             Drugstore    Dumpling Restaurant   
102              Burger Joint            Kids Store         Sandwich Place   

     10th Most Common Venue  
1               Escape Room  
2                Shoe Store  
3           Airport Service  
4        Chinese Restaurant  
7        Chinese Restaurant  
..                      ...  
98                Drugstore  
99                    Hotel  
100        Recording Studio  
101       Electronics Store  
102  Thrift / Vintage Store  

[88 rows x 12 columns]

#### Cluster 3

In [77]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0   Martial Arts School           Yoga Studio   

          3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45  Eastern European Restaurant   Distribution Center               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45      Doner Restaurant            Donut Shop             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
45   Dumpling Restaurant      Electronics Store

#### Cluster 4

In [78]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0            Print Shop   Dumpling Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11        Discount Store   Distribution Center               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11      Doner Restaurant            Donut Shop             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
11           Yoga Studio     Dim Sum Restaurant

#### Cluster 5

In [79]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant   Dumpling Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6        Discount Store   Distribution Center               Dog Run   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6      Doner Restaurant            Donut Shop             Drugstore   

         9th Most Common Venue   10th Most Common Venue  
6  Eastern European Restaurant  Health & Beauty Service